### DISCUSSION TITLE
`executeOperation`: Support `TypedDocumentNode` generic type

### DISCUSSION THREAD

**Discussion by jangerhofer** [apollographql/apollo-server#7579] [Ideas]
Howdy, @glasser?

As @tobiasdiez (among others) [commented](https://github.com/apollographql/apollo-server/issues/5464#issuecomment-903188901), `ApolloServer.executeOperation` support for the  [`TypedDocumentNode` format](https://github.com/dotansimha/graphql-typed-document-node)[^1] would advance type-safety for all who use [`graphql-code-generator`](https://github.com/dotansimha/graphql-code-generator) alongside integration tests.  I see that you both made `executeOperation` generic over the [`TypedQueryDocumentNode` type](https://github.com/apollographql/apollo-server/blob/3d0d4b4b263e52bb36839221d9103f02ba1d6288/docs/source/api/apollo-server.mdx?plain=1#L641) from `graphql-js` in #6960 and considered `TypedDocumentNode` support around the same time.  If I am not mistaken, such support never landed!

I am eager to take a stab at making `apollo-server` play nicely with the `graphql-code-generator` standard in tests, but want to check 

(a) that you're [still](https://github.com/apollographql/apollo-server/pull/5847#issuecomment-974558245) [in favor](https://github.com/apollographql/apollo-server/pull/6960#issue-1388165676) of this change 😄 and

(b) whether you have any particular opinion about how to approach implementation.

Are you willing to offer guidance?

[^1]: Specifically, support for inferring the `variables` parameter & return type  of `executeOperation` based on the type of the GQL document passed as the `query` parameter.

**Comment by trevor-scheer** [comment]
@jangerhofer go for it, that'd be great! Happy to offer guidance as needed.

Presumably adding `TypedDocumentNode` to the union type for `request.query` (the argument to `executeOperation`) should work. As far as I can tell it should just be a compatible addition with no runtime changes needed.

Some tests which show the type correctness of using a `TypedDocumentNode` would be great.

**Reply by jangerhofer** [reply]
Awesome -- thanks @trevor-scheer!  I'll circle back with a diff soon.
